src/cycle.rs

```diff
@@ -1,92 +1,16 @@
-pub struct Cycle {
-    participants: CycleParticipants,
-}
-
-// We want `Cycle`` to be thin
-pub(crate) type CycleParticipants = Arc<Box<[DatabaseKeyIndex]>>;
-
-impl Cycle {
-    pub(crate) fn new(participants: CycleParticipants) -> Self {
-        Self { participants }
-    }
+/// Should only be relevant in case of a badly configured cycle recovery.
+pub const MAX_ITERATIONS: u32 = 200;
 
-    /// True if two `Cycle` values represent the same cycle.
-    pub(crate) fn is(&self, cycle: &Cycle) -> bool {
-        Arc::ptr_eq(&self.participants, &cycle.participants)
-    }
+/// Return value from a cycle recovery function.
+#[derive(Debug)]
+pub enum CycleRecoveryAction<T> {
+    /// Iterate the cycle again to look for a fixpoint.
+    Iterate,
 
-    pub(crate) fn throw(self) -> ! {
-        tracing::debug!("throwing cycle {:?}", self);
-        std::panic::resume_unwind(Box::new(self))
-    }
-
-    pub(crate) fn catch<T>(execute: impl FnOnce() -> T) -> Result<T, Cycle> {
-        match std::panic::catch_unwind(AssertUnwindSafe(execute)) {
-            Ok(v) => Ok(v),
-            Err(err) => match err.downcast::<Cycle>() {
-                Ok(cycle) => Err(*cycle),
-                Err(other) => std::panic::resume_unwind(other),
-            },
-        }
-    }
-
-    /// Iterate over the [`DatabaseKeyIndex`] for each query participating
-    /// in the cycle. The start point of this iteration within the cycle
-    /// is arbitrary but deterministic, but the ordering is otherwise determined
-    /// by the execution.
-    pub fn participant_keys(&self) -> impl Iterator<Item = DatabaseKeyIndex> + '_ {
-        self.participants.iter().copied()
-    }
-
-    /// Returns a vector with the debug information for
-    /// all the participants in the cycle.
-    pub fn all_participants(&self, _db: &dyn Database) -> Vec<DatabaseKeyIndex> {
-        self.participant_keys().collect()
-    }
-
-    /// Returns a vector with the debug information for
-    /// those participants in the cycle that lacked recovery
-    /// information.
-    pub fn unexpected_participants(&self, db: &dyn Database) -> Vec<DatabaseKeyIndex> {
-        self.participant_keys()
-            .filter(|&d| d.cycle_recovery_strategy(db) == CycleRecoveryStrategy::Panic)
-            .collect()
-    }
-}
-
-impl std::fmt::Debug for Cycle {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        crate::attach::with_attached_database(|db| {
-            f.debug_struct("UnexpectedCycle")
-                .field("all_participants", &self.all_participants(db))
-                .field("unexpected_participants", &self.unexpected_participants(db))
-                .finish()
-        })
-        .unwrap_or_else(|| {
-            f.debug_struct("Cycle")
-                .field("participants", &self.participants)
-                .finish()
-        })
-    }
+    /// Cut off iteration and use the given result value for this query.
+    Fallback(T),
```

Comment:
In the current implementation, the initial function will still be called. If query A(0) is executing, it calls B(0), which in turn calls A(0), the second call to A(0) will call the initial function to get an initial value to return to B(0), which will compute its provisional value and return that to A(0), which will then call the cycle recovery function. If that cycle recovery function immediately returns `Fallback`, then that fallback value will be used for A(0). (I think we also need to then iterate the cycle one more time in order to populate a correct value for B(0) based on the fallback value for A(0), but that's not currently implemented.)

It seems like it could be possible to instead call the cycle recovery function first thing when we encounter the cycle, and allow it to return `Fallback` immediately, before the initial function is ever called. That seems worth experimenting with.

Reply:
A key difference in this PR from the current cycle handling is that in the current cycle handling _all_ cycle participants that have a cycle fallback receive a fallback value; this ensures determinism regardless of query order, but it means that you can't have some participants in the cycle compute a real value based on a fallback for some other cycle participant.

In this PR, if your recovery function falls back to some value, that fallback only applies to the individual query, and the rest of the cycle will then resolve normally using that fallback value.

Reply:
I see, thanks, that explanation made this very clear!